runtime/jitserver_launcher/jitserver.c

```diff
@@ -0,0 +1,308 @@
+#endif	
+
+_end:
+	return buffer;
+}
+
+static int32_t
+isJava8(char *jdkOrJreHome)
+{
+	int32_t rc = 0;
+	int32_t pathLen = strlen(jdkOrJreHome);
+	char *lastLevel = NULL;
+	int32_t restoreTrailingSep = 0;
+
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "isJava8> jdkOrJreHome: %s\n", jdkOrJreHome);
+#endif
+
+	/* Consume any trailing '/'. We should restore it before exit. */
+	if (jdkOrJreHome[pathLen - 1] == DIR_SEPARATOR_CHAR) {
+		jdkOrJreHome[pathLen - 1] = '\0';
+		restoreTrailingSep = 1;
+	}
+
+	lastLevel = strrchr(jdkOrJreHome, DIR_SEPARATOR_CHAR);
+	if (NULL == lastLevel) {
+		/* Did not find DIR_SEPARATOR_CHAR => we are at the root level and there is no jre directory */
+		rc = 0;
+	} else if (!strcmp("jre", lastLevel + 1)) {
+		/* Found jre directory => we are Java 8 */
+		rc = 1;
+	} else {
+		rc = 0;
+	}
+
+	if (1 == restoreTrailingSep) {
+		jdkOrJreHome[pathLen - 1] = DIR_SEPARATOR_CHAR;
+	}
+
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "isJava8> returning %d\n", rc);
+#endif
+	return rc;
+}
+
+static void 
+truncateOneLevel(char *path)
+{
+	char *lastLevel = NULL;
+	int32_t pathLen = strlen(path);
+
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "truncateOneLevel> on entry, path: %s\n", path);
+#endif
+
+	/* Consume any trailing '/' */
+	if (path[pathLen - 1] == DIR_SEPARATOR_CHAR) {
+		path[pathLen - 1] = '\0';
+	}
+
+	/* Now consume the last directory if present */	
+	lastLevel = strrchr(path, DIR_SEPARATOR_CHAR);
+	if (NULL != lastLevel) {
+		*lastLevel = '\0';	
+	}
+
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "truncateOneLevel> on exit, path: %s\n", path);
+#endif
+}
+
+static int32_t 
+append(char *buffer, int32_t bufLen, char *str)
+{
+	if (strlen(buffer) + strlen(str) + 1 > bufLen) {
+		return -1;
+	} else {
+		strcat(buffer, str);
+		return 0;
+	}
+}
+
+static char * 
+getJvmLibPath(void)
+{
+	char *jvmLibPath = NULL;
+	char *selfPath = NULL;
+	int32_t addArchDir = 0;
+	int32_t rc = 0;
+
+	selfPath = getSelfPath();
+	if (NULL == selfPath) {
+		goto _end;
+	}
+	jvmLibPath = malloc(J9_MAX_PATH);
+	if (NULL == jvmLibPath) {
+		fprintf(stderr, "Failed to allocate memory for jvm library\n");
+		goto _end;
+	}
+	strcpy(jvmLibPath, selfPath);
+	free(selfPath);
+
+	/* selfPath is <java_home>/bin/jitserver for Java 9+
+	 * or <java_home>/jre/bin/jitserver for Java 8.
+	 * We need to find libjvm.so at <java_home>/lib/j9vm/libjvm.so for Java 9+
+	 * or <java_home>/jre/lib/amd64/j9vm/libjvm.so for Java 8.
+	 * To reach <java_home> we need to remove two entries from the path.
+	 */
+	truncateOneLevel(jvmLibPath);
+	truncateOneLevel(jvmLibPath);
+
+	addArchDir = isJava8(jvmLibPath);
+	if (1 == addArchDir) {
+		rc = append(jvmLibPath, J9_MAX_PATH, "/lib/" OPENJ9_ARCH_DIR "/j9vm/libjvm.so");
+	} else {
+		rc = append(jvmLibPath, J9_MAX_PATH, "/lib/j9vm/libjvm.so");
+	}
+	if (0 != rc) {
+		fprintf(stderr, "Path exceeded maximum length\n");
```

Comment:
It would be useful to the user to know the maximum value for the path length. Let's add it to this message